test/integration/network/services_test.go

```diff
@@ -0,0 +1,458 @@
+			}
+			etcd := framework.SharedEtcd()
+			// cleanup the registry storage
+			defer registry.CleanupStorage()
+			// start a kube-apiserver
+			server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{
+				"--service-cluster-ip-range", tc.serviceCIDR,
+				"--advertise-address", "10.0.0.1",
+				"--feature-gates", fmt.Sprintf("IPv6DualStack=%t", tc.dualStack),
+			}, etcd)
+			// create a client
+			client, err := kubernetes.NewForConfig(server.ClientConfig)
+			if err != nil {
+				t.Errorf("error creating client: %v", err)
+			}
+
+			// verify client is working
+			if err := wait.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
+				_, err = client.CoreV1().Endpoints("default").Get("kubernetes", metav1.GetOptions{})
+				if err != nil {
+					t.Logf("error fetching endpoints: %v", err)
+					return false, nil
+				}
+				return true, nil
+			}); err != nil {
+				t.Errorf("server without enabled endpoints failed to register: %v", err)
+			}
+
+			// Create a service setting the IPFamily and check that it's created with the correct IP family
+			svc := createService("svc-ipfamily-set", &ipFamily, v1.ServiceTypeClusterIP)
+			if _, err := client.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+				t.Errorf("unexpected error text: %v", err)
+			}
+
+			// Create a service without IPFamily and check that it's created with the default IP family
+			svc = createService("svc-ipfamily-nil", nil, v1.ServiceTypeClusterIP)
+			if _, err := client.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+				t.Errorf("unexpected error text: %v", err)
+			}
+
+			// obtain the services and check the Spec.IPFamily is correct
+			svcs, err := client.CoreV1().Services(metav1.NamespaceDefault).List(metav1.ListOptions{})
+			if err != nil {
+				t.Fatalf("unexpected success, and error getting the services: %v", err)
+			}
+			for _, s := range svcs.Items {
+				// single family Spec.IPFamily must be nil
+				if !tc.dualStack {
+					if s.Spec.IPFamily != nil {
+						t.Fatalf("expected service IP family nil: %s ", s.Spec.ClusterIP)
+					}
+				} else if err = validateServiceAndClusterIPFamily(&s, ipFamily); err != nil {
+					t.Fatalf("unexpected service IP family %s: %v", s.Spec.ClusterIP, err)
+				}
+			}
+
+			// Create a service with the non-default IPfamily, it must fail on non-dualstack clusters
+			svc = createService("svc-ipfamily-other", &otherIPFamily, v1.ServiceTypeClusterIP)
+			_, err = client.CoreV1().Services(metav1.NamespaceDefault).Create(svc)
+			if !tc.dualStack && err == nil {
+				t.Errorf("unexpected success creating service: %v", svc)
+			}
+			if tc.dualStack && err != nil {
+				t.Errorf("unexpected error text: %v", err)
+			}
+		})
+	}
+}
+
+// TestTransitionService ClusterIP -> Headless -> External -> clear ClusterIP -> ClusterIP
+// TestServicesIPFamily tests the IP family field on DualStack clusters
+func TestTransitionService(t *testing.T) {
```

Comment:
need to go deeper into this test but this is the output I'm getting
```
    --- FAIL: TestTransitionService/DualStack_enabled_and_IPv6_IPv4 (4.80s)
        testserver.go:181: runtime-config=map[api/all:true]
        testserver.go:182: Starting kube-apiserver on port 52840...
        testserver.go:198: Waiting for /healthz to be ok...
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.externalName: Required value, spec.ipFamily: Required value]
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.clusterIP: Invalid value: "": field is immutable, spec.ipFamily: Invalid value: "null": field is immutable, spec.ipFamily: Required value]
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.clusterIP: Invalid value: "": field is immutable, spec.ipFamily: Invalid value: "null": field is immutable, spec.ipFamily: Required value]
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.clusterIP: Invalid value: "": field is immutable, spec.ipFamily: Invalid value: "null": field is immutable, spec.ipFamily: Required value]
```

Reply:
the first error is because you are trying to set type=externalname type without specifying the required externalName field.

the other errors might be related to the bug @smarterclayton is trying to fix, I'm not sure